### PR TITLE
feat: filter changes using go list output

### PR DESCRIPTION
**Pull request by kuisathaverat**
## What does this PR do?

It extracts the PATH dependencies from the list of Go vendor modules

## Why is it important?

This approach is automatic when you add new dependencies, also, it is more accurate filtering better the changes.

## Related issues

Relates https://github.com/elastic/beats/issues/16953

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -850,93 +857,88 @@ def isChangedXPackCode(patterns) {
...
+  env.BUILD_GENERATOR = isChangedOSSCode(getVendorPatterns('generator'))
+}
 
-  env.GO_VERSION = readFile(".go-version").trim()
+/**
+  This method grab the dependencies of a Go module and transform them on regexp
+*/
+def getVendorPatterns(beatName){
+  def output = ""
+  docker.image("golang:${GO_VERSION}").inside{
```
Comment (v1v) [inline Jenkinsfile:935]: NIT: delegate the docker pull with a retry/sleep to minimise the environmental issues when pulling from internet.

```suggestion
  def dockerImage = "golang:${GO_VERSION}"
  retry(2) {
    sleep 2
    sh "docker pull ${dockerImage}"
  }
  docker.image("${dockerImage}").inside{
```

File: Jenkinsfile [hunk 2]
```diff
@@ -662,7 +663,9 @@ def withBeatsEnv(boolean archive, Closure body) {
       sh(label: "Install docker-compose ${DOCKER_COMPOSE_VERSION}", script: ".ci/scripts/install-docker-compose.sh")
       sh(label: "Install Mage", script: "make mage")
       try {
-        body()
+        if(!params.dry_run){
```
Comment (v1v) [inline Jenkinsfile:666]: Can the `dry_run` move to the beginning of the methods: `withBeatsEnvWin ` and `withBeatsEnv`? Or is it required to run the sh/bat steps?

File: Jenkinsfile (hunk 1 above)
Comment (kuisathaverat) [reply Jenkinsfile:935]: sounds like a new step for the library

Comment (kuisathaverat) [reply Jenkinsfile:935]: this image is in the packer cache so, in theory, it should not take it from internet

File: Jenkinsfile (hunk 2 above)
Comment (kuisathaverat) [reply Jenkinsfile:666]: in that case we do not test the installer.

**Comment by kuisathaverat** [comment]
@urso @jsoriano @andrewkroh 
The k8s test is only triggered if you change a file in `deploy/kubernetes/` for me looks wrong, it means that your code changes will not be tested on k8s. I'd said that merges to master should trigger those test also.

**Comment by jsoriano** [comment]
> @urso @jsoriano @andrewkroh
> The k8s test is only triggered if you change a file in `deploy/kubernetes/` for me looks wrong, it means that your code changes will not be tested on k8s. I'd said that merges to master should trigger those test also.

Agree, changes in other places can also affect deployments in kubernetes.

**Comment by urso** [comment]
I see I'm a little late to the discussion, but I wonder how effective this change will be.

Can we document what exactly `getVendorPatterns` will return? Just from reading the code it is far from obvious. Does it contain the package names or the vendor paths?

Given that we always add the regular expressions `^libbeat/.*` and `^vendor/.*` I wonder what the exact effect of `goVendorPatterns` will be.
If `getVendorPatterns` creates regular expressions, would it make sense to anchor those with `^`?

Running the provided script locally I get paths like this:

```
github.com/gofrs/uuid/.*
github.com/gogo/protobuf/proto/.*
github.com/gogo/protobuf/sortkeys/.*
github.com/golang/protobuf/proto/.*
github.com/golang/protobuf/ptypes/.*
github.com/golang/protobuf/ptypes/any/.*
github.com/golang/protobuf/ptypes/duration/.*
github.com/golang/protobuf/ptypes/timestamp/.*
github.com/golang/snappy/.*
```

These are clearly dependencies that should be in `vendor`. But the path is incorrect.

Only a small set of packages are prefixed with `vendor`:

```
vendor/golang.org/x/crypto/chacha20/.*
vendor/golang.org/x/crypto/chacha20poly1305/.*
vendor/golang.org/x/crypto/cryptobyte/.*
vendor/golang.org/x/crypto/cryptobyte/asn1/.*
vendor/golang.org/x/crypto/curve25519/.*
vendor/golang.org/x/crypto/hkdf/.*
vendor/golang.org/x/crypto/internal/subtle/.*
vendor/golang.org/x/crypto/poly1305/.*
vendor/golang.org/x/net/dns/dnsmessage/.*
vendor/golang.org/x/net/http/httpguts/.*
vendor/golang.org/x/net/http/httpproxy/.*
vendor/golang.org/x/net/http2/hpack/.*
vendor/golang.org/x/net/idna/.*
vendor/golang.org/x/sys/cpu/.*
vendor/golang.org/x/text/secure/bidirule/.*
vendor/golang.org/x/text/transform/.*
vendor/golang.org/x/text/unicode/bidi/.*
vendor/golang.org/x/text/unicode/norm/.*
```

No idea why these packages are prefixed with vendor, but the output actually includes these packages twice:

```
golang.org/x/text/transform/.*
golang.org/x/text/unicode/bidi/.*
golang.org/x/text/unicode/norm/.*
vendor/golang.org/x/text/transform/.*
vendor/golang.org/x/text/unicode/bidi/.*
vendor/golang.org/x/text/unicode/norm/.*
```

**Comment by kuisathaverat** [comment]
>Can we document what exactly getVendorPatterns will return? Just from reading the code it is far from obvious. Does it contain the package names or the vendor paths?

It returns the output of `go list -mod=vendor` adding `.*` at the end of each line. It also removes the string`github.com/elastic/beats/v7` to make the beats paths match.

>These are clearly dependencies that should be in vendor. But the path is incorrect.

because not all packages have `vendor` on the beginning the regexp is generated without '^'

>No idea why these packages are prefixed with vendor, but the output actually includes these packages twice:

we can filter them in a follow-up

**Pull request merged**